use asyncio/aiohttp and drop database/queueing

jluebbe opened the pull request:
This is a quick hack to rip out worker threads and queuing via sqlite, replacing them with asyncio and aiohttpd. Each PDURunner now has an `asyncio.Lock` to serialize the `loop.run_in_executor` calls, which then run the actual driver code in a asyncio thread pool. This only implements blocking mode (see #75).

I've tried to keep the general structure close to the original code to make it easier to follow the changes. Only --drive mode and the http listener work right now, and I've only tested the `localcmdline` driver.

Further work:
- [ ] replace urlparse and listener.parse_http with the aiohttp mechanisms
- [x] port the tcp listener
- [x] clean up signal handling and shutdown

Possible future work:
- [ ] show available drivers in the `/` path
- [ ] print incremental feedback during request processing (waiting, retrying, ...)

jluebbe commented:
@mattface What do you think about the general approach? I'd like to get your feedback before pushing this further.

mattface commented:
I've tried this out and with a quick patch to remove the --dbfile argument, it functions but currently you cant close the daemon as it doesnt seem to handle the signals.
I think we can continue with this idea as removing the database requirement was always intended.

jluebbe commented:
I've converted the TCPListener to asyncio as well and implemented the suggestions by @cdleonard.

There's still some cleanup that could be done, but doing that in this PR would turn it into a complete rewrite of the affected files, making review more difficult. So I'd prefer to keep that for later.

atline commented:
`Blocking mode` is something guys have wait for long time, maybe too long. So, if this MR could be taken as priority will really be great...

hubifab commented:
> I've tried this out and with a quick patch to remove the --dbfile argument, it functions but currently you cant close the daemon as it doesnt seem to handle the signals. I think we can continue with this idea as removing the database requirement was always intended.

@mattface I just tried signalling pdudaemon with SIGTERM and Ctrl+c, both seemed to work fine now. Is anything still open to do? Looking forward to have this in the main branch! :smiley:
I have a tiny feature I would like to add on top and don't want to make things messy.
Thanks for all your time!

mattface commented:
Hi all,

Apologies for the slow reply. Thanks for the work on this, it's a big improvement.
I've pulled this locally, rebased on main and have run the basic tests that github actions would run and have a couple of issues

First is a simple one, the python version was too old:
```
future: <Task finished coro=<PDUDaemon.shutdown() done, defined at /usr/local/lib/python3.7/dist-packages/pdudaemon-0.1-py3.7.egg/pdudaemon/__init__.py:111> exception=TypeError("shutdown() got an unexpected keyword argument 'cancel_futures'")>
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/pdudaemon-0.1-py3.7.egg/pdudaemon/__init__.py", line 116, in shutdown
    await runner.shutdown()
  File "/usr/local/lib/python3.7/dist-packages/pdudaemon-0.1-py3.7.egg/pdudaemon/pdurunner.py", line 43, in shutdown
    self.executor.shutdown(wait=True, cancel_futures=True)
TypeError: shutdown() got an unexpected keyword argument 'cancel_futures'
```

so this patch cures it:
```
diff --git a/Dockerfile.dockerhub b/Dockerfile.dockerhub
index 9614990..b86bb55 100644
--- a/Dockerfile.dockerhub
+++ b/Dockerfile.dockerhub
@@ -1,4 +1,4 @@
-FROM debian:buster
+FROM debian:bullseye
 
 ARG HTTP_PROXY
 ENV http_proxy ${HTTP_PROXY}
@@ -16,6 +16,7 @@ libudev-dev \
 libusb-1.0-0-dev \
 pkg-config \
 psmisc \
+python3.9 \
 python3-pip \
 python3-setuptools \
 python3-usb \
```


But the error I havent had chance to debug is:
```
2022-08-15 19:52:12,940 - pdud                           - INFO Starting the listener
2022-08-15 19:52:12,940 - pdud.tcp                       - INFO listening on 0.0.0.0:16421
/usr/local/lib/python3.9/dist-packages/pdudaemon-0.1-py3.9.egg/pdudaemon/tcplistener.py:68: RuntimeWarning: coroutine 'wait_for' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
2022-08-15 19:52:15,218 - pdud.tcp                       - DEBUG <class 'TypeError'>
2022-08-15 19:52:15,218 - pdud.tcp                       - DEBUG 'coroutine' object is not subscriptable
Error sending command! localhost replied: 'coroutine' object is not subscriptable
```

which prevents the TCP listener from accepting any requests.

I do wonder if anyone cares about the TCP listener as it really should be considered legacy by now, but if it's a small fix I'd prefer to keep it working.

I think first step is please @jluebbe can you rebase to main and we can work from there

jluebbe commented:
> Apologies for the slow reply. Thanks for the work on this, it's a big improvement. I've pulled this locally, rebased on main and have run the basic tests that github actions would run and have a couple of issues
> 
> First is a simple one, the python version was too old:
> 
> ```
> future: <Task finished coro=<PDUDaemon.shutdown() done, defined at /usr/local/lib/python3.7/dist-packages/pdudaemon-0.1-py3.7.egg/pdudaemon/__init__.py:111> exception=TypeError("shutdown() got an unexpected keyword argument 'cancel_futures'")>
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.7/dist-packages/pdudaemon-0.1-py3.7.egg/pdudaemon/__init__.py", line 116, in shutdown
>     await runner.shutdown()
>   File "/usr/local/lib/python3.7/dist-packages/pdudaemon-0.1-py3.7.egg/pdudaemon/pdurunner.py", line 43, in shutdown
>     self.executor.shutdown(wait=True, cancel_futures=True)
> TypeError: shutdown() got an unexpected keyword argument 'cancel_futures'
> ```

We could probably support 3.7 as well, if that's still useful. Note that it will be EOL on [2023-06-27](https://www.python.org/downloads/), so I'd tend to drop support for it.

> so this patch cures it:
> 
> ```
> diff --git a/Dockerfile.dockerhub b/Dockerfile.dockerhub
> index 9614990..b86bb55 100644
> --- a/Dockerfile.dockerhub
> +++ b/Dockerfile.dockerhub
> @@ -1,4 +1,4 @@
> -FROM debian:buster
> +FROM debian:bullseye
>  
>  ARG HTTP_PROXY
>  ENV http_proxy ${HTTP_PROXY}
> @@ -16,6 +16,7 @@ libudev-dev \
>  libusb-1.0-0-dev \
>  pkg-config \
>  psmisc \
> +python3.9 \
>  python3-pip \
>  python3-setuptools \
>  python3-usb \
> ```

The buster to bullseye change should be enough, as 3.9 is the the only supported version there.

> But the error I havent had chance to debug is:
> 
> ```
> 2022-08-15 19:52:12,940 - pdud                           - INFO Starting the listener
> 2022-08-15 19:52:12,940 - pdud.tcp                       - INFO listening on 0.0.0.0:16421
> /usr/local/lib/python3.9/dist-packages/pdudaemon-0.1-py3.9.egg/pdudaemon/tcplistener.py:68: RuntimeWarning: coroutine 'wait_for' was never awaited
> RuntimeWarning: Enable tracemalloc to get the object allocation traceback
> 2022-08-15 19:52:15,218 - pdud.tcp                       - DEBUG <class 'TypeError'>
> 2022-08-15 19:52:15,218 - pdud.tcp                       - DEBUG 'coroutine' object is not subscriptable
> Error sending command! localhost replied: 'coroutine' object is not subscriptable
> ```
> 
> which prevents the TCP listener from accepting any requests.
> 
> I do wonder if anyone cares about the TCP listener as it really should be considered legacy by now, but if it's a small fix I'd prefer to keep it working.
> 
> I think first step is please @jluebbe can you rebase to main and we can work from there

I've rebased (there were no conflicts) and fixed the tcplistener issue (we need to await first and then get element 0).
